src/opnsense/mvc/app/library/OPNsense/Mvc/Router.php

```diff
@@ -0,0 +1,181 @@
+                if (is_file($expected_filename)) {
+                    return  $new_namespace;
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Route a request
+     * @param string $uri
+     * @param array $default list of routing defaults (controller, acount)
+     * @return Response to be rendered
+     * @throws ClassNotFoundException
+     * @throws MethodNotFoundException
+     * @throws ParameterMismatchException
+     * @throws InvalidUriException
+     * @throws ReflectionException when invoke fails
+     */
+    public function routeRequest(string $uri, array $defaults = []) : Response
+    {
+        $path = parse_url($uri)['path'];
+
+        if (!str_starts_with($path, $this->prefix)) {
+            throw new InvalidUriException("Invalid route path: " . $uri);
+        }
+
+        // extract target (base)namespace, controller and action
+        $targetAndParameters = $this->parsePath(substr($path, strlen($this->prefix)), $defaults);
+
+        $controller = $targetAndParameters['controller'];
+        // resolve full namespace (including vendor) when we know which controller to access.
+        $namespace = $this->resolveNamespace($targetAndParameters['namespace'], $controller);
+        $action = $targetAndParameters['action'];
+        $parameters = $targetAndParameters['parameters'];
+
+        if ($action === null || $controller === null || $namespace === null) {
+            throw new InvalidUriException("Invalid route path, no action, controller, and / or namespace: " . $uri);
+        }
+
+        $dispatcher = new Dispatcher($namespace, $controller, $action, $parameters);
+
+        return $this->performRequest($dispatcher);
+    }
+
+    /**
+     * @param Dispatcher $dispatcher request dispatcher
+     * @return Response object
+     * @throws ClassNotFoundException
+     * @throws MethodNotFoundException
+     * @throws ParameterMismatchException
+     * @throws ReflectionException when invoke fails
+     */
+    private function performRequest(Dispatcher $dispatcher) : Response
+    {
+        $session = new Session();
+        $request = new Request();
+        $response = new Response();
+
+        $dispatcher->dispatch($request, $response, $session);
+        return $response;
+    }
+
+
+    /**
+     * @param string $path path to extract
+     * @param array $default list of routing defaults (controller, acount)
+     * @return array containing expected controller action
+     */
+    private function parsePath(string $path, array $defaults) : array {
+        $pathElements = explode("/", rtrim($path, '/'));
+        $result = [
+            "namespace" => null,
+            "controller" => null,
+            "action" => null,
+            "parameters" => []
+        ];
+        foreach ($defaults as $key => $val) {
+            $result[$key] = $val;
+        }
+
+        $count = 0;
+        foreach($pathElements as $element){
+            if($count == 0) {
```

Comment:
the foreach-if-count construct is a little over-engineered. maybe this could be flattened using `array_shift()` in the future as well

Reply:
some style issues in that block which are better fixed now, I'll fix those, use the `$idx`  from the array itself and be done with it :)